File: lc_soln/0853_car_fleet/solution.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    int carFleet(int target, vector<int>& position, vector<int>& speed) {
        int n = position.size();
        std::vector<std::pair<int, int>> pos_and_spd;
        pos_and_spd.reserve(n);
        for (int i = 0; i < position.size(); ++i) {
            pos_and_spd.emplace_back(position[i], speed[i]);
        }
        std::sort(pos_and_spd.begin(), pos_and_spd.end(),
                  [](std::pair<int, int>& a, std::pair<int, int>& b) { return a.first > b.first; });
        std::vector<double> result;
        for (const std::pair<int, int>& p : pos_and_spd) {
            double time = static_cast<double>(target - p.first) / static_cast<double>(p.second);
            if (!result.empty() && time <= result.back()) {
                continue;
            }
            result.push_back(time);
        }
        return result.size();
    }
};

File: lc_soln/0853_car_fleet/solution.rs
struct Solution;

impl Solution {
    pub fn car_fleet(target: i32, position: Vec<i32>, speed: Vec<i32>) -> i32 {
        let mut pos_and_spd: Vec<(i32, i32)> = position.into_iter().zip(speed.into_iter()).collect();
        pos_and_spd.sort_unstable_by_key(|(pos, _)| -pos);
        let mut stack = vec![];
        for (pos, spd) in pos_and_spd.into_iter() {
            let time = (target - pos) as f64 / spd as f64;
            if let Some(&last) = stack.last() && time <= last {
                continue;
            }
            stack.push(time);
        }
        stack.len() as i32
    }
}

fn main() {
}
